pkg/service/did/ion.go

```diff
@@ -0,0 +1,289 @@
+	}
+	if err = h.keyStore.StoreKey(ctx, *updateStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion update private key")
+	}
+
+	recoveryStoreRequest, err := keyToStoreRequest(ionDID.ID()+"#"+recoverKeySuffix, ionDID.GetRecoveryPrivateKey(), ionDID.ID())
+	if err != nil {
+		return nil, errors.Wrap(err, "converting recovery private key to store request")
+	}
+	if err = h.keyStore.StoreKey(ctx, *recoveryStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion recovery private key")
+	}
+
+	keyStoreRequest, err := keyToStoreRequest(ionDID.ID()+"#"+keyID, *privKeyJWK, ionDID.ID())
+	if err != nil {
+		return nil, errors.Wrap(err, "converting private key to store request")
+	}
+	if err = h.keyStore.StoreKey(ctx, *keyStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion private key")
+	}
+
+	privKeyBytes, err := crypto.PrivKeyToBytes(privKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "converting private key to bytes")
+	}
+	return &CreateDIDResponse{
+		DID:              didDoc,
+		PrivateKeyBase58: base58.Encode(privKeyBytes),
+		KeyType:          request.KeyType,
+	}, nil
+}
+
+func keyToStoreRequest(kid string, privateKeyJWK crypto.PrivateKeyJWK, controller string) (*keystore.StoreKeyRequest, error) {
+	privateKey, err := privateKeyJWK.ToPrivateKey()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting private private key from JWK")
+	}
+	keyType, err := crypto.GetKeyTypeFromPrivateKey(privateKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting private key type from private privateKeyJWK")
+	}
+	// convert to a serialized format
+	privateKeyBytes, err := crypto.PrivKeyToBytes(privateKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not encode private key as base58 string")
+	}
+	privateKeyBase58 := base58.Encode(privateKeyBytes)
+	return &keystore.StoreKeyRequest{
+		ID:               kid,
+		Type:             keyType,
+		Controller:       controller,
+		PrivateKeyBase58: privateKeyBase58,
+	}, nil
+}
+
+func (h *ionHandler) GetDID(ctx context.Context, request GetDIDRequest) (*GetDIDResponse, error) {
+	id := request.ID
+
+	// TODO(gabe) as we are fully custodying ION DIDs this is fine; as we move to a more decentralized model we will
+	//  need to either remove local storage or treat it as a cache with a TTL
+
+	// first check if the DID is in the storage
+	gotDID := new(ionStoredDID)
+	err := h.storage.GetDID(ctx, id, gotDID)
+	if err == nil {
+		return &GetDIDResponse{DID: gotDID.DID}, nil
+	}
+	logrus.WithError(err).Warnf("error getting DID from storage: %s", id)
+
+	// if not, resolve it from the network
+	resolved, err := h.resolver.Resolve(ctx, id, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "resolving DID from network")
+	}
+	return &GetDIDResponse{DID: resolved.Document}, nil
+}
+
+// GetDIDs returns all DIDs we have in storage for ION, it is not feasible to get all DIDs from the network
```

Comment:
there is no Ion batch call functionality ?

Reply:
looking at the api - [no](https://github.com/decentralized-identity/ion/blob/master/src/core.ts) - this is an improvement we can make